app/routes.py

```diff
@@ -1,2 +1,326 @@
+import os
+
+from flask import abort
+
+# this method loads the values from the .env file which the os module can then read 
+load_dotenv()
+
+# instantiating blueprints for task and goal
+tasks_bp = Blueprint("tasks", __name__, url_prefix="/tasks")
+goals_bp = Blueprint("goals", __name__, url_prefix="/goals")
+
+
+# HELPER FUNCTIONS:
+#=============================================================================
+
+def slack_post_to_task_notifications_channel(text):
+    """
+    inputs: text (string), which is the message to be posted
+    outputs: a message posted to the task-notifications channel in Slack
+    """
+    # os.environ.get() basically conjures up the value referenced by SLACKBOT_TOKEN in the .env file
+    # requests.post() takes an optional headers argument that can be passed in as a dictionary like this
+    # According to the Slack API documentation for the chat.postMessage endpoint, the token should be passed as an HTTP Authorization header 
+    post_headers = {"Authorization": os.environ.get("SLACKBOT_TOKEN")}
+    
+    # channel is the other required parameter for the chat.postMessage endpoint; was probably redundant to make the channel ID secret but I did anyway 
+    # text is the actual string message to be posted in Slack
+    post_data = {
+        "channel": os.environ.get("TASK_NOTIFICATIONS_CHANNEL_ID"),
+        "text": text
+    }
+
+    # this is the code that induces the bot to make the post by sending a POST request 
+    # The required parameters for this endpoint are the token (post_headers) and channel (post_data)
+    # interesting to note that PATCH or POST or PUT all work here as verbs, which is unexpected given the Slack documentation for this endpoint
+    requests.post('https://slack.com/api/chat.postMessage', headers=post_headers, data=post_data)
+
+
+def valid_id_or_400(input_id):
+    """
+    input: an input id 
+    output: a 400 status code if the input ID is not able to be converted to an integer  
+    """
+    # tries to convert the input_id to an integer
+    try:
+        return int(input_id)
+    # if a ValueError is raised, an abort with 400 gets triggered 
+    except ValueError:
+        abort(400,{"message": f"ID {input_id} must be an integer", "success": False})
+
+
+# TASK ENDPOINTS:
+#=============================================================================
+
+# this is a decorator that USES the tasks_bp Blueprint to define an endpoint (/<task_id>) and accepted method(s) (GET)  
+# This function will execute whenever a request that matches the decorator is received. The name of this function doesn't affect how requests are routed to this method   
+@tasks_bp.route("/<task_id>", methods=["GET"], strict_slashes=False)
+def get_single_task(task_id):
+
+    # ❓ Would it make sense to bundle checking for valid task id and querying a task for that id into one instance method or helper function - since I'm doing that for every endpoint
```

Comment:
Great question! The question I would ask to answer that question is: "When I call this function, do I always query for a task as well?" And in this code, I see the answer is yes! So I think that would be a great little improvement.

Reply:
Okay, cool - The reason I held off is that I think I'm a little fuzzy still on when it makes sense to bundle repeat code like that vs. employing SRP guidelines? I was worried that putting both of those things (checking for input validity, returning 400 if invalid - If valid, querying an instance of something, returning 404 if it can't find it) was too much stuff for a helper function to be doing?